src/System.Memory/src/System/Buffers/SequenceReader.cs

```diff
@@ -0,0 +1,373 @@
+                AdvanceToNextSpan(count);
+            }
+        }
+
+        /// <summary>
+        /// Unchecked helper to avoid unnecessary checks where you know count is valid.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal void AdvanceCurrentSpan(long count)
+        {
+            Debug.Assert(count >= 0);
+
+            Consumed += count;
+            CurrentSpanIndex += (int)count;
+            if (CurrentSpanIndex >= CurrentSpan.Length)
+                GetNextSpan();
+        }
+
+        /// <summary>
+        /// Only call this helper if you know that you are advancing in the current span
+        /// with valid count and there is no need to fetch the next one.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal void AdvanceWithinSpan(long count)
+        {
+            Debug.Assert(count >= 0);
+
+            Consumed += count;
+            CurrentSpanIndex += (int)count;
+
+            Debug.Assert(CurrentSpanIndex < CurrentSpan.Length);
+        }
+
+        private void AdvanceToNextSpan(long count)
+        {
+            if (count < 0)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.count);
+            }
+
+            Consumed += count;
+            while (_moreData)
+            {
+                int remaining = CurrentSpan.Length - CurrentSpanIndex;
+
+                if (remaining > count)
+                {
+                    CurrentSpanIndex += (int)count;
+                    count = 0;
+                    break;
+                }
+
+                CurrentSpanIndex += remaining;
+                count -= remaining;
+                Debug.Assert(count >= 0);
+
+                GetNextSpan();
+
+                if (count == 0)
+                    break;
+            }
+
+            if (count != 0)
+            {
+                // Not enough space left- adjust for where we actually ended and throw
+                Consumed -= count;
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.count);
+            }
+        }
+
+        /// <summary>
+        /// Peek forward the number of positions in <paramref name="copyBuffer"/> (at most), copying into
+        /// <paramref name="copyBuffer"/> if needed.
+        /// </summary>
+        /// <param name="copyBuffer">
+        /// Temporary buffer to copy into if there isn't a contiguous span within the existing data to return.
+        /// Also describes the maximum count of positions to peek.
+        /// </param>
+        /// <returns>
+        /// Span over the peeked data. The length may be shorter than <paramref name="copyBuffer"/> if there
+        /// is not enough data left to fill the requested length.
+        /// </returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public ReadOnlySpan<T> Peek(Span<T> copyBuffer)
```

Comment:
Was this API used when you plumbed this into Kestrel?  If not, I'm wondering if it should be left out for now.  I'm not really sure when I'd choose to use this as a consumer vs just using `UnreadSpan` myself and only if it was of insufficient length then allocating a copy buffer (either stack or heap) and passing that to some `CopyTo`-like method.

Reply:
I guess `TryCopyTo` is ok for this scenario.